packages/apputils-extension/src/index.ts

```diff
@@ -611,10 +611,34 @@ const utilityCommands: JupyterFrontEndPlugin<void> = {
  * The default HTML sanitizer.
  */
 const sanitizer: JupyterFrontEndPlugin<ISanitizer> = {
-  id: '@jupyter/apputils-extension:sanitizer',
+  id: '@jupyterlab/apputils-extension:sanitizer',
   autoStart: true,
   provides: ISanitizer,
-  activate: () => {
+  requires: [ISettingRegistry],
+  activate: (app: JupyterFrontEnd, settings: ISettingRegistry) => {
+    const loadSetting = (setting: ISettingRegistry.ISettings): void => {
+      const schemesArray = setting.get('additionalSchemes')
+        .composite as Array<string>;
+
+      schemesArray.forEach(scheme => {
+        if (defaultSanitizer.addScheme) {
+          defaultSanitizer.addScheme(scheme);
```

Comment:
We need something slightly different here because if the user changes the schemes from the settings this will append the full new list. That means, if a scheme is preserved, it will be added an additional time and if it is removed, it won't actually be as the previous values are kept.

We could rather define a `setSchemes`. The trouble is that since this is public API of `ISanitizer` another plugin can mess up with the settings.

So I would go for implementing a `setSchemes` and to not add anything to the public API (aka the interface `ISanitizer` in the file `token.ts`)

Reply:
Ah okay fair enough. Sorry I'm a bit new to JuypterLab development. Would implementing `setSchemes` mean creating a new class that extends `Sanitizer` so we wouldn't have to modify the interface?

Reply:
??

Reply:
This should be possible without creation new class. The interface specifies a subset of class API, if a class has a method which is not in the interface it is ok

Reply:
If I try to add `setSchemes` in `Sanitizer` at `https://github.com/jupyterlab/jupyterlab/blob/master/packages/apputils/src/sanitizer.ts` and call it in `https://github.com/jupyterlab/jupyterlab/blob/master/packages/apputils-extension/src/index.ts` I get a:  

```ts
Property 'setScheme' does not exist on type 'ISanitizer'.ts(2339)
```

(sorry if I'm missing something super obvious)

Reply:
No, it's me who missed that `defaultSanitizer` is casted to `ISanitizer` in:

https://github.com/jupyterlab/jupyterlab/blob/a059144fd7946f1f953ffec8d9b56baf283d5cf7/packages/apputils/src/sanitizer.ts#L962

Some ways I could quickly think of:
- add a new interface with only `setSchemes` method, say `ISchemaConfigurable` and mark `defaultSanitizer` as implementing a union of `ISchemaConfigurable` and `ISanitizer`, or
- cast the `defaultSanitizer` to `Sanitizer` type with `as` (I think we should avoid it if possible)
- add an alias for `defaultSanitizer` in `apputils/src/sanitizer.ts` which would not be casted to `ISanitizer`

Maybe @fcollonval has a better idea though.

Reply:
Ah okay. Thanks for the quick reply. I'll look into implementing something on the back off those points and `await` @fcollonval response if there are any other suitable ideas. Thanks! @krassowski

Reply:
For this PR targeting `master`, we should get ride of `defaultSanitizer` in favor of the token based sanitizer - otherwise we will get inconsistency for people using a different plugin or (as it will happen with this PR) non-default settings.

You will instantiate a `new Sanitizer` (after removing the default one) in this `activate` function. So that you get access the full API (including the `setSchemes`) but the token consumer should not rely on it as it is not exposed in the interface.

---

Unfortunately the default sanitizer is use in the code in two places:

1. in the `RenderMimeRegistry`:

https://github.com/jupyterlab/jupyterlab/blob/60491d8185d86597d3f1882b277824635b60fc93/packages/rendermime/src/registry.ts#L41

There you can simply replace it with a new instance of the sanitizer

```patch
- options.sanitizer ?? defaultSanitizer;
+ options.sanitizer ?? new Sanitizer();
```

1. in the completer

https://github.com/jupyterlab/jupyterlab/blob/60491d8185d86597d3f1882b277824635b60fc93/packages/completer/src/widget.ts#L1013

That one is more tricky because you gonna need to provide the sanitizer as option to the completer constructor and pass it from the token in `updateCompleter` calls in three extensions: `console-extension`, `notebook-extension` and `fileeditor-extension` (look for `updateCompleter` to find them).

Reply:
Thanks @fcollonval I'll try have a crack at it.

Reply:
@fcollonval Had a crack at it. Wasn't exactly sure how to pass it in the `updateCompleter` calls but was able to add it in the constructor to the `Completer`.

Reply:
Leaving this here. Let me know if there is anything else I need to do. Thanks.